server/src/Icepeak/Server/WebsocketServer/HandleClientMultiSubscription.hs

```diff
@@ -0,0 +1,140 @@
+
+import qualified Data.Aeson as Aeson
+import qualified Data.ByteString.Lazy as LBS
+import qualified Network.WebSockets as WS
+
+import Icepeak.Server.Core (Core, coreClients, withCoreMetrics, SubscriberState (SubscriberStateNew))
+import qualified Icepeak.Server.Metrics as Metrics
+import qualified Icepeak.Server.Subscription as Subscription
+
+-- * Client handling
+
+newUUID :: IO UUID
+newUUID = randomIO
+
+data ClientPayload
+  = Subscribe   { subscribePath :: [Text], token :: Text }
+  | UnSubscribe { unsubscribePath :: [Text] }
+
+instance Aeson.FromJSON ClientPayload where
+  parseJSON = undefined
+
+data ClientPayloadParsed
+  = ClientPayloadSuccessfulParse ClientPayload
+  | ClientPayloadFailedParse
+
+clientPayloadFromByteString :: LBS.ByteString -> ClientPayloadParsed
+clientPayloadFromByteString byteString = 
+  case Aeson.eitherDecode byteString of
+      (Left _jsonDecodeError) -> ClientPayloadFailedParse
+      (Right clientPayload)   -> ClientPayloadSuccessfulParse clientPayload
+
+clientPayloadFromDataMessage :: WS.DataMessage -> ClientPayloadParsed
+clientPayloadFromDataMessage (WS.Text encodedUtf8ByteString _mbDecodedCache) =
+  clientPayloadFromByteString encodedUtf8ByteString
+clientPayloadFromDataMessage (WS.Binary _) = ClientPayloadFailedParse
+  
+authorisePathSubscription :: [Text] -> Text -> IO Bool
+authorisePathSubscription = undefined
+
+handleClient :: WS.Connection -> Core -> IO ()
+handleClient conn core = do
+  uuid <- newUUID
+
+  isDirtyMVar <- newMVar ()
+  subscribedPathsMVar <- newMVar (HashMap.empty :: HashMap [Text] (MVar Value))
+
+  let
+    serverStateMVar = coreClients core
+
+    onConnect = withCoreMetrics core Metrics.incrementSubscribers
+
+    onDisconnect = do
+      paths <-  HashMap.keys <$> takeMVar subscribedPathsMVar
+      for_ paths
+        (\path -> modifyMVar_ serverStateMVar
+          (pure . Subscription.unsubscribe path uuid))
+      withCoreMetrics core Metrics.decrementSubscribers
+
+    onPayload (ClientPayloadSuccessfulParse (Subscribe newPath pathToken)) = do
+      isAuthorised <- authorisePathSubscription newPath pathToken
+
+      case isAuthorised of
+        False -> undefined
+        True -> do
+          pathValueMVar <- newEmptyMVar
+          modifyMVar_ subscribedPathsMVar
+            (pure . HashMap.insert newPath pathValueMVar)
+          modifyMVar_ serverStateMVar
+            (pure . Subscription.subscribe newPath uuid
+              (SubscriberStateNew (pathValueMVar, isDirtyMVar)))
+
+    onPayload (ClientPayloadSuccessfulParse (UnSubscribe unsubPath)) = do
+      modifyMVar_ serverStateMVar
+        (pure . Subscription.unsubscribe unsubPath uuid)
+      modifyMVar_ subscribedPathsMVar
+        (pure . HashMap.delete unsubPath)
+
+    onPayload ClientPayloadFailedParse = pure ()
+
+    takeMVarNewValues = do
+      takeMVar isDirtyMVar
+      valueMVars <- HashMap.elems <$> readMVar subscribedPathsMVar
+      catMaybes <$> mapM tryTakeMVar valueMVars
```

Comment:
1. Yes, its worth to ponder about a bit more. Can think about if a less redundant prone implementation. And yes, sounds like a nice thing to do some signalling of some threshold amount of subscriptions for client implementation errors. And perhaps it can be of a similar nature as to the duplicate subscriptions warning/error.
2. I have not thought about this yet. It does feel like this logic to control sending can be placed inside `takeMVarNewValues`